net/sync/gpoddernet/src/main/java/de/danoeh/antennapod/net/sync/nextcloud/NextcloudSyncService.java

```diff
@@ -0,0 +1,224 @@
+            throw new SyncServiceException(e);
+        } catch (Exception e) {
+            e.printStackTrace();
+            throw new SyncServiceException(e);
+        }
+    }
+
+    private HashMap<String, List<String>> getHeaderWithUserAgent() {
+        HashMap<String, List<String>> header = new HashMap<>();
+        header.put("User-Agent", Collections.singletonList(userAgent));
+        return header;
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.O)
+    @Override
+    public UploadChangesResponse uploadSubscriptionChanges(List<String> addedFeeds, List<String> removedFeeds) {
+        try {
+            HashMap<String, List<String>> header = getHeaderWithUserAgent();
+            header.put("Content-Type", Collections.singletonList("application/json"));
+
+            final JSONObject requestObject = new JSONObject();
+            requestObject.put("add", new JSONArray(addedFeeds));
+            requestObject.put("remove", new JSONArray(removedFeeds));
+
+            NextcloudRequest nextcloudRequest = new NextcloudRequest.Builder()
+                    .setMethod("POST")
+                    .setUrl(Uri.encode("/index.php/apps/gpoddersync/subscription_change/create", "/"))
+                    .setHeader(header)
+                    .setRequestBody(requestObject.toString())
+                    .build();
+
+            this.nextcloudApi.performNetworkRequest(nextcloudRequest).close();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        return new GpodnetUploadChangesResponse(now().getEpochSecond(), new HashMap<>());
+    }
+
+    @Override
+    public EpisodeActionChanges getEpisodeActionChanges(long timestamp) throws SyncServiceException {
+        try {
+            String uri = "/index.php/apps/gpoddersync/episode_action";
+            HashMap<String, String> parameters = new HashMap<>();
+            parameters.put("since", String.valueOf(timestamp));
+            NextcloudRequest nextcloudRequest = new NextcloudRequest.Builder()
+                    .setMethod("GET")
+                    .setUrl(Uri.encode(uri, "/"))
+                    .setParameter(parameters)
+                    .build();
+
+            InputStream inputStream = this.nextcloudApi.performNetworkRequest(nextcloudRequest);
+            String responseString = ResponseMapper.convertStreamToString(inputStream);
+            inputStream.close();
+            JSONObject json = new JSONObject(responseString);
+            return ResponseMapper.readEpisodeActionsFromJsonObject(json);
+        } catch (URISyntaxException e) {
+            e.printStackTrace();
+            throw new IllegalStateException(e);
+        } catch (JSONException | MalformedURLException e) {
+            e.printStackTrace();
+            throw new SyncServiceException(e);
+        } catch (Exception e) {
+            e.printStackTrace();
+            throw new SyncServiceException(e);
+        }
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.O)
```

Comment:
What happens when using this on an older Android version?

Reply:
it probably would break. I replaced this with "legacy"  approach